File: apps/android/app/src/main/java/com/karibuhealth/app/ui/components/AiReviewHelpers.kt
package com.karibuhealth.app.ui.components

import com.karibuhealth.app.data.remote.dto.AiReviewSuggestionDto
import com.karibuhealth.app.ui.dictation.NoteSection
import com.karibuhealth.app.ui.util.stripCitationArtifacts

data class DictationIncorporate(
    val suggestionId: String,
    val section: NoteSection,
    val prefill: String,
    val openLabPicker: Boolean = false,
    val openRxPicker: Boolean = false,
)

fun dedupeAiReviewSuggestions(suggestions: List<AiReviewSuggestionDto>): List<AiReviewSuggestionDto> {
    val pending = suggestions.filter { it.clinicianResponse == null }
    if (pending.size <= 1) return pending

    val kept = mutableListOf<AiReviewSuggestionDto>()
    for (candidate in pending) {
        val normQ = normalizeQuestion(candidate.question)
        val duplicate = kept.any { existing ->
            existing.suggestionType == candidate.suggestionType &&
                questionSimilarity(normalizeQuestion(existing.question), normQ) >= 0.72
        }
        if (!duplicate) kept.add(candidate)
    }
    return kept
}

fun incorporationFor(suggestion: AiReviewSuggestionDto): DictationIncorporate {
    val section = when (suggestion.suggestionType) {
        "ask_lab" -> NoteSection.TestsOrdered
        "ask_dx" -> NoteSection.Diagnosis
        "ask_med" -> NoteSection.Medications
        "ask_history" -> NoteSection.Hpi
        "ask_red_flag" -> NoteSection.PhysicalExam
        else -> NoteSection.AssessmentPlan
    }
    val prefill = "Consider: ${suggestion.question.trim()}"
    return DictationIncorporate(
        suggestionId = suggestion.id,
        section = section,
        prefill = prefill,
        openLabPicker = suggestion.suggestionType == "ask_lab",
        openRxPicker = suggestion.suggestionType == "ask_med",
    )
}

fun displayAiReasoning(reasoning: String): String = stripCitationArtifacts(reasoning)

private fun normalizeQuestion(text: String): String =
    text.lowercase()
        .replace(Regex("""[^\w\s]"""), " ")
        .replace(Regex("""\s+"""), " ")
        .trim()

private fun questionSimilarity(a: String, b: String): Double {
    if (a.isBlank() || b.isBlank()) return 0.0
    if (a == b) return 1.0
    val wordsA = a.split(" ").filter { it.length > 2 }.toSet()
    val wordsB = b.split(" ").filter { it.length > 2 }.toSet()
    if (wordsA.isEmpty() || wordsB.isEmpty()) return 0.0
    val intersection = wordsA.intersect(wordsB).size
    val union = wordsA.union(wordsB).size
    return intersection.toDouble() / union.toDouble()
}

File: apps/web/src/lib/ai-review-helpers.ts
import type { NoteSectionKey } from '@/lib/clinical-note-sections'

/** Timeline-tier AI notes (docs/ai-clinical-assist.md). Interruptive alerts use visit_critical_alerts. */

export type AiReviewSuggestionLike = {
  clinician_response: string | null
  display_tier?: string | null
  phase?: string | null
}

export type VisitAiReviewSuggestion = {
  id: string
  suggestion_type: string
  question: string
  reasoning: string
  citation_ids: number[]
  confidence: 'high' | 'medium' | 'low'
  clinician_response: 'considered_proceeded' | 'reopened_note' | 'dismissed' | null
  citations?: Array<{
    id: number
    document_title: string
    document_slug: string
    section: string | null
    section_anchor: string | null
  }>
}

export const ACTIONABLE_AI_SUGGESTION_TYPES = new Set([
  'ask_lab',
  'ask_med',
  'ask_dx',
  'ask_history',
  'ask_red_flag',
])

export type DictationIncorporate = {
  suggestionId: string
  section: NoteSectionKey
  prefill: string
  openLabPicker: boolean
  openRxPicker: boolean
}

export function filterTimelineAiNotes<T extends AiReviewSuggestionLike>(
  suggestions: T[],
): T[] {
  return suggestions
    .filter(
      (s) =>
        !s.clinician_response &&
        (s.display_tier ?? 'timeline') === 'timeline' &&
        (s.phase === 'draft' || s.phase === 'lab' || s.phase == null),
    )
    .slice(0, 3)
}

/** Map an AI suggestion to editor prefill (mirrors Android incorporationFor). */
export function incorporationFor(suggestion: Pick<VisitAiReviewSuggestion, 'id' | 'suggestion_type' | 'question'>): DictationIncorporate {
  const section: NoteSectionKey = (() => {
    switch (suggestion.suggestion_type) {
      case 'ask_lab':
        return 'TestsOrdered'
      case 'ask_dx':
        return 'Diagnosis'
      case 'ask_med':
        return 'Medications'
      case 'ask_history':
        return 'Hpi'
      case 'ask_red_flag':
        return 'PhysicalExam'
      default:
        return 'AssessmentPlan'
    }
  })()

  return {
    suggestionId: suggestion.id,
    section,
    prefill: `Consider: ${suggestion.question.trim()}`,
    openLabPicker: suggestion.suggestion_type === 'ask_lab',
    openRxPicker: suggestion.suggestion_type === 'ask_med',
  }
}

export function stripCitationArtifacts(reasoning: string): string {
  return reasoning.replace(/\[chunk_id=\d+]/g, '').replace(/\s{2,}/g, ' ').trim()
}
